In [225]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [226]:
df = sns.load_dataset('iris')

In [227]:
df.head()

sepal_length  sepal_width  petal_length  petal_width species
0           5.1          3.5           1.4          0.2  setosa
1           4.9          3.0           1.4          0.2  setosa
2           4.7          3.2           1.3          0.2  setosa
3           4.6          3.1           1.5          0.2  setosa
4           5.0          3.6           1.4          0.2  setosa

In [228]:
df['species'].unique()

array(['setosa', 'versicolor', 'virginica'], dtype=object)

In [229]:
X  = df.drop('species', axis=1)
y = df['species']

In [230]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [231]:
from sklearn.neighbors import KNeighborsClassifier

In [232]:
modelknn = KNeighborsClassifier(n_neighbors=2)
modelknn.fit(X_train, y_train)

KNeighborsClassifier(n_neighbors=2)

In [233]:
modelknn.score(X_test, y_test)

1.0

In [234]:
from sklearn.svm import SVC

In [235]:
model_svm= SVC(C=10, kernel='linear', gamma='auto')

In [236]:
model_svm.fit(X_train, y_train)

SVC(C=10, gamma='auto', kernel='linear')

In [237]:
model_svm.score(X_test, y_test)

0.9666666666666667

As we seen above as we change the parameters the score of the model changes but checking all the possible values manually is not possible so we use some methods like Grid Search CV, Random Search CV

In [238]:
from sklearn.model_selection import GridSearchCV

In [239]:
model_svm= SVC(gamma='auto')

In [240]:
classifier = GridSearchCV((model_svm),{
    'C':[1,10,20,30],
    'kernel':['linear','rbf'],
    'gamma':['auto','scale']
},cv=5,return_train_score=False)

# Similar above we can add multiple params

In [241]:
# As we are using Cross Validation here we use whole data to fit the model and not split one

classifier.fit(X,y)

GridSearchCV(cv=5, estimator=SVC(gamma='auto'),
             param_grid={'C': [1, 10, 20, 30], 'gamma': ['auto', 'scale'],
                         'kernel': ['linear', 'rbf']})

In [242]:
classifier.cv_results_

{'mean_fit_time': array([0.01328993, 0.01094127, 0.01513133, 0.00952177, 0.00953126,
        0.00729809, 0.00729628, 0.00879698, 0.00542088, 0.00598426,
        0.00672607, 0.00509634, 0.00495863, 0.00564189, 0.00422497,
        0.00404477]),
 'std_fit_time': array([0.00233191, 0.00250051, 0.00344182, 0.00254989, 0.00431071,
        0.00036415, 0.00225585, 0.00150749, 0.00088306, 0.00108096,
        0.00104974, 0.00082064, 0.00113011, 0.00133291, 0.00096995,
        0.00088184]),
 'mean_score_time': array([0.00869727, 0.01040106, 0.00645251, 0.00662112, 0.00561786,
        0.00549488, 0.00671525, 0.00515218, 0.00737553, 0.0048697 ,
        0.00662074, 0.00352278, 0.00339913, 0.0036386 , 0.00313444,
        0.0034379 ]),
 'std_score_time': array([0.00414332, 0.00181724, 0.0013759 , 0.00068204, 0.00096686,
        0.00146344, 0.00094443, 0.00080948, 0.00331058, 0.00167781,
        0.00127007, 0.00072053, 0.00091989, 0.00081631, 0.00124109,
        0.00097937]),
 'param_C': masked_array(d

In [243]:
results = pd.DataFrame(classifier.cv_results_)

In [244]:
results

mean_fit_time  std_fit_time  mean_score_time  std_score_time param_C  \
0        0.013290      0.002332         0.008697        0.004143       1   
1        0.010941      0.002501         0.010401        0.001817       1   
2        0.015131      0.003442         0.006453        0.001376       1   
3        0.009522      0.002550         0.006621        0.000682       1   
4        0.009531      0.004311         0.005618        0.000967      10   
5        0.007298      0.000364         0.005495        0.001463      10   
6        0.007296      0.002256         0.006715        0.000944      10   
7        0.008797      0.001507         0.005152        0.000809      10   
8        0.005421      0.000883         0.007376        0.003311      20   
9        0.005984      0.001081         0.004870        0.001678      20   
10       0.006726      0.001050         0.006621        0.001270      20   
11       0.005096      0.000821         0.003523        0.000721      20   
12       0.004959      0.001130         0.003399        0.000920      30   
13       0.005642      0.001333         0.003639        0.000816      30   
14       0.004225      0.000970         0.003134        0.001241      30   
15       0.004045      0.000882         0.003438        0.000979      30   

   param_gamma param_kernel                                           params  \
0         auto       linear    {'C': 1, 'gamma': 'auto', 'kernel': 'linear'}   
1         auto          rbf       {'C': 1, 'gamma': 'auto', 'kernel': 'rbf'}   
2        scale       linear   {'C': 1, 'gamma': 'scale', 'kernel': 'linear'}   
3        scale          rbf      {'C': 1, 'gamma': 'scale', 'kernel': 'rbf'}   
4         auto       linear   {'C': 10, 'gamma': 'auto', 'kernel': 'linear'}   
5         auto          rbf      {'C': 10, 'gamma': 'auto', 'kernel': 'rbf'}   
6        scale       linear  {'C': 10, 'gamma': 'scale', 'kernel': 'linear'}   
7        scale          rbf     {'C': 10, 'gamma': 'scale', 'kernel': 'rbf'}   
8         auto       linear   {'C': 20, 'gamma': 'auto', 'kernel': 'linear'}   
9         auto          rbf      {'C': 20, 'gamma': 'auto', 'kernel': 'rbf'}   
10       scale       linear  {'C': 20, 'gamma': 'scale', 'kernel': 'linear'}   
11       scale          rbf     {'C': 20, 'gamma': 'scale', 'kernel': 'rbf'}   
12        auto       linear   {'C': 30, 'gamma': 'auto', 'kernel': 'linear'}   
13        auto          rbf      {'C': 30, 'gamma': 'auto', 'kernel': 'rbf'}   
14       scale       linear  {'C': 30, 'gamma': 'scale', 'kernel': 'linear'}   
15       scale          rbf     {'C': 30, 'gamma': 'scale', 'kernel': 'rbf'}   

    split0_test_score  split1_test_score  split2_test_score  \
0            0.966667           1.000000           0.966667   
1            0.966667           1.000000           0.966667   
2            0.966667           1.000000           0.966667   
3            0.966667           0.966667           0.966667   
4            1.000000           1.000000           0.900000   
5            0.966667           1.000000           0.966667   
6            1.000000           1.000000           0.900000   
7            0.966667           1.000000           0.966667   
8            1.000000           1.000000           0.900000   
9            0.966667           1.000000           0.900000   
10           1.000000           1.000000           0.900000   
11           0.966667           1.000000           0.966667   
12           1.000000           1.000000           0.900000   
13           0.966667           1.000000           0.900000   
14           1.000000           1.000000           0.900000   
15           0.966667           1.000000           0.933333   

    split3_test_score  split4_test_score  mean_test_score  std_test_score  \
0            0.966667                1.0         0.980000        0.016330   
1            0.966667                1.0         0.980000        0.016330   
2            0.966667                1.0       

From above we can check which paramter combo is giving best accuracy means mean_test_score

In [245]:
results[['param_C','param_kernel','mean_test_score']]

param_C param_kernel  mean_test_score
0        1       linear         0.980000
1        1          rbf         0.980000
2        1       linear         0.980000
3        1          rbf         0.966667
4       10       linear         0.973333
5       10          rbf         0.980000
6       10       linear         0.973333
7       10          rbf         0.980000
8       20       linear         0.966667
9       20          rbf         0.966667
10      20       linear         0.966667
11      20          rbf         0.980000
12      30       linear         0.960000
13      30          rbf         0.960000
14      30       linear         0.960000
15      30          rbf         0.973333

As from above we get that 	C=1	kernel = linear	 give highest accuracy as 0.980000

In [246]:
model_knn = KNeighborsClassifier(weights='distance')

In [247]:
knn_classifier = GridSearchCV((model_knn),{
    'n_neighbors':[1,3,2,5,10],
    'metric':['euclidean','manhattan','minkowski'],
    'p':[1,2],
    
},cv=5,return_train_score=False)

# Similar above we can add multiple params

In [248]:
knn_classifier.fit(X,y)

GridSearchCV(cv=5, estimator=KNeighborsClassifier(weights='distance'),
             param_grid={'metric': ['euclidean', 'manhattan', 'minkowski'],
                         'n_neighbors': [1, 3, 2, 5, 10], 'p': [1, 2]})

In [249]:
knn_results = pd.DataFrame(knn_classifier.cv_results_)

In [250]:
knn_results

mean_fit_time  std_fit_time  mean_score_time  std_score_time param_metric  \
0        0.007196      0.000958         0.007270        0.002739    euclidean   
1        0.004507      0.000464         0.006507        0.001923    euclidean   
2        0.005095      0.001269         0.005576        0.000693    euclidean   
3        0.004358      0.000562         0.005720        0.001484    euclidean   
4        0.004200      0.002042         0.004118        0.000768    euclidean   
5        0.003485      0.001038         0.004486        0.000589    euclidean   
6        0.003444      0.000436         0.029555        0.052917    euclidean   
7        0.002472      0.000610         1.371420        2.735100    euclidean   
8        0.002199      0.000947         0.003949        0.001449    euclidean   
9        0.002559      0.001012         0.002519        0.000777    euclidean   
10       0.002435      0.001077         0.004402        0.004222    manhattan   
11       0.002556      0.000946         0.002321        0.000408    manhattan   
12       0.002001      0.000636         0.003211        0.001044    manhattan   
13       0.002547      0.000547         0.002201        0.000486    manhattan   
14       0.004722      0.002262         0.008166        0.001873    manhattan   
15       0.005632      0.000573         0.005820        0.000938    manhattan   
16       0.003416      0.001255         0.005301        0.000663    manhattan   
17       0.003712      0.001054         0.004191        0.000524    manhattan   
18       0.003229      0.000492         0.004059        0.001476    manhattan   
19       0.003382      0.000682         0.003570        0.001072    manhattan   
20       0.003987      0.001991         0.003142        0.001088    minkowski   
21       0.002918      0.000829         0.003449        0.000886    minkowski   
22       0.002738      0.000770         0.003090        0.000739    minkowski   
23       0.002246      0.000482         0.002503        0.000861    minkowski   
24       0.002295      0.000400         0.002249        0.000380    minkowski   
25       0.002851      0.000743         0.002915        0.000300    minkowski   
26       0.001968      0.000970         0.002198        0.000402    minkowski   
27       0.002610      0.000617         0.002566        0.000566    minkowski   
28       0.002209      0.000393         0.002232        0.000389    minkowski   
29       0.001798      0.000401         0.002331        0.000579    minkowski   

   param_n_neighbors param_p  \
0                  1       1   
1                  1       2   
2                  3       1   
3                  3       2   
4                  2       1   
5                  2       2   
6                  5       1   
7                  5       2   
8                 10       1   
9                 10       2   
10                 1       1   
11                 1       2   
12                 3       1   
13                 3       2   
14                 2       1   
15                 2       2   
16                 5       1   
17                 5       2   
18                10       1   
19                10       2   
20                 1       1   
21                 1       2   
22                 3       1   
23                 3       2   
24                 2       1   
25                 2       2   
26                 5       1   
27                 5       2   
28                10       1   
29                10       2   

                                               params  split0_test_score  \
0   {'metric': 'euclidean', 'n_neighbors': 1, 'p': 1}           0.966667   
1   {'metric': 'euclidean', 'n_neighbors': 1, 'p': 2}           0.966667   
2   {'metric': 'euclidean', 'n_neighbors': 3, 'p': 1}           0.966667   
3   {'metric': 'euclidean', 'n_neighbors': 3, 'p': 2}           0.966667   
4   {'metric': 'euclidean', 'n_neighbors': 2, 'p': 1}           0.966667   
5   {'metric': 'euclidean', 'n_neighbors': 

In [251]:
knn_results[['param_n_neighbors','param_metric','mean_test_score']]

param_n_neighbors param_metric  mean_test_score
0                  1    euclidean         0.960000
1                  1    euclidean         0.960000
2                  3    euclidean         0.966667
3                  3    euclidean         0.966667
4                  2    euclidean         0.960000
5                  2    euclidean         0.960000
6                  5    euclidean         0.966667
7                  5    euclidean         0.966667
8                 10    euclidean         0.986667
9                 10    euclidean         0.986667
10                 1    manhattan         0.960000
11                 1    manhattan         0.960000
12                 3    manhattan         0.960000
13                 3    manhattan         0.960000
14                 2    manhattan         0.960000
15                 2    manhattan         0.960000
16                 5    manhattan         0.960000
17                 5    manhattan         0.960000
18                10    manhattan         0.973333
19                10    manhattan         0.973333
20                 1    minkowski         0.960000
21                 1    minkowski         0.960000
22                 3    minkowski         0.960000
23                 3    minkowski         0.966667
24                 2    minkowski         0.960000
25                 2    minkowski         0.960000
26                 5    minkowski         0.960000
27                 5    minkowski         0.966667
28                10    minkowski         0.973333
29                10    minkowski         0.986667

**Random Seach CV**

In [252]:
from sklearn.model_selection  import RandomizedSearchCV

In [253]:
classifier_r = RandomizedSearchCV((model_svm),{
    'C':[1,10,20,30],
    'kernel':['linear','rbf']
},n_iter=30,cv=5,return_train_score=False)


In [254]:
classifier_r.fit(X,y)

c:\Users\manes\anaconda3\Lib\site-packages\sklearn\model_selection\_search.py:307: UserWarning: The total space of parameters 8 is smaller than n_iter=30. Running 8 iterations. For exhaustive searches, use GridSearchCV.
  warnings.warn(


RandomizedSearchCV(cv=5, estimator=SVC(gamma='auto'), n_iter=30,
                   param_distributions={'C': [1, 10, 20, 30],
                                        'kernel': ['linear', 'rbf']})

In [255]:
results_rscv = pd.DataFrame(classifier_r.cv_results_)
results_rscv[['param_C','param_kernel','mean_test_score']]

param_C param_kernel  mean_test_score
0       1       linear         0.980000
1       1          rbf         0.980000
2      10       linear         0.973333
3      10          rbf         0.980000
4      20       linear         0.966667
5      20          rbf         0.966667
6      30       linear         0.960000
7      30          rbf         0.960000